docs/content/v0.20/architecture.md

```diff
@@ -1,36 +1,479 @@
+follows:
+
+1.  Informer invokes the event handler AddFunc or UpdateFunc
+
+2.  The event handler does some filtering (for example, to ignore update
+    events where the object content is not modified) and then creates a
+    key to store a reference to the object in the work queue. The key
+    contains the GVK key used to retrieve a reference to the informer
+    and lister for that kind of object, and a namespace/name key to
+    retrieve the actual object. Storing the key in the work queue is a
+    common pattern in client-go as the object may have changed in the
+    cache (which is kept updated by the reflector) by the time a worker
+    gets a copy from the work queue. Workers should always receive the
+    key and use it to retrieve the object from the cache.
+
+3.  A worker pulls a key from the work queue, and then does the
+    following processing:
+
+    -  Uses the GVK key to get the reference to the lister for the
+        object
+    -  Gets the object from the lister cache using the NamespacedName of 
+       the object.
+    -  Gets the lister for Placement objects and list all placements
+    -  Iterates on all placements, and for each placement:
+        - evaluates the placement label selector vs. object labels to
+          check if there is a match.
+        - If there is a match, uses the lister for ManagedCluster and
+          find the matching clusters using the label selector expression
+          for clusters.
+        - If there are matching clusters, add the names of the cluster
+          to a hashmap setting the name of the cluster as a key. Cluster
+          groups from different placements are merged together.
+    -  If there are no matching clusters, the worker returns without
+        actions and is ready to process other events from the queue.
+    -  If there are matching clusters:
+        - Wraps the object into a ManifestWork
+        - Adds a label for each matched (Placement, WDS) to the ManifestWork that 
+          is used to track the placement that caused the object to be 
+          delivered to one or more clusters. The label contains both the 
+          placement name (note that placement is cluster-scoped) and the WDS name. 
+          This way, when a placement is deleted or updated it is possible to 
+          locate the associated ManifestWorks for deletion or label removal 
+          (if more than one label is present, as other placements may “own” the object).
+        - For each cluster:
+          -  Sets the manifest namespace == name of the cluster
+          -  Uses the client connected to the ITS to do a server-side
+              apply of the manifest into the namespace.
+        - Worker returns and is ready to pick other keys from the queue.
+
+There are other event flows, based on the object GVK and type of event.
+The following sections broadly describe these flows:
+
+#### Object Deleted 
+
+When an object is deleted from the WDS, the handler’s *DeleteFunc* is
```

Comment:
What if a worker dequeues a key that refers to an object that is _now_ deleted (but wasn't when the event handler ran)?

You may say "no problem, the fact that the object is _now_ deleted implies that there was a following event handler invocation that inserted a key with a copy of the object".  There are two failures of the sequencing assumptions there. One is that the two keys are not equal and thus not combined in the workqueue and thus may be processed concurrently in different workers, thus not serialized. And even if all the work wrt a given object _were_ serialized, the first attempt to handle the non-deletion key might suffer a transient failure causing the key to be requeued for later retry, possibly after the work on the deletion key.

Reply:
The worker tries to retrieve the object if it was not deleted when the event handler ran:

https://github.com/kubestellar/kubestellar/blob/e516d75cc05dcfe16b041d8616381ceaaaea4f1a/pkg/placement/controller.go#L390-L402

As for the concurrency, we can discuss in https://github.com/kubestellar/kubestellar/issues/1674

Reply:
This doc should say somewhere what happens when the worker is processing a `DeletedObject`-free key and gets `errors.IsNotFound`.

Reply:
Agree, not an error, normal behavior in Kube.

As for what happens, I am updating the section about create/update (step 3) with this text:

```

Reply:
So the doc issue has been resolved. There remains the code issue of reporting a non-error as an error.